diff -u src/df/df.c src/df/df.c
--- src/df/df.c	2026-04-01 11:45:13.902968963 -0500
+++ src/df/df.c	2026-04-01 16:38:33.517117230 -0500
@@ -45,12 +45,13 @@
 #if 0
 static char sccsid[] = "@(#)df.c	8.7 (Berkeley) 4/2/94";
 #else
 __RCSID("$NetBSD: df.c,v 1.101.2.1 2023/12/18 14:17:42 martin Exp $");
 #endif
 #endif /* not lint */
 
 #include <sys/param.h>
 #include <sys/stat.h>
+#include <sys/statvfs.h>
 #include <sys/mount.h>
 
 #include <assert.h>
@@ -64,20 +65,55 @@
 #include <stdlib.h>
 #include <string.h>
 #include <unistd.h>
+#include <mntent.h>
 #include <util.h>
 
+#include "compat.h"
+
+struct mntinfo {
+    unsigned long int f_frsize;
+    fsfilcnt_t f_favail;
+    unsigned long int f_namemax;
+    unsigned int f_type;    
+    unsigned long int f_fsid;
+    char *f_mntfromname;          /* mnt_fsname from getmntent */
+    char *f_mntfromlabel;	  /* THIS WONT BE USE IT */
+    char *f_mntonname;            /* mnt_dir from getmntent */
+    char *f_fstypename;           /* mnt_fsname from getmntent */
+    char *f_opts;                 /* mnt_opts from getmntent */
+    unsigned long f_bsize;        /* f_bsize from statvfs */
+    fsblkcnt_t f_blocks;          /* f_blocks from statvfs */
+    fsblkcnt_t f_bfree;           /* f_bfree from statvfs */
+    fsblkcnt_t f_bavail;          /* f_bavail from statvfs */
+    fsfilcnt_t f_files;           /* f_files from statvfs */
+    fsfilcnt_t f_ffree;           /* f_ffree from statvfs */
+    unsigned long f_flag;         /* f_flag from statvfs */
+    dev_t f_dev;                  /* st_dev from stat */
+    unsigned int f_selected;      /* used internally here only */
+    uid_t f_owner; 		  /* THIS WONT BE USE IT */
+    uint64_t f_syncwrites;	  /* THIS WONT BE USE IT */
+    uint64_t f_asyncwrites;	  /* THIS WONT BE USE IT */
+    fsblkcnt_t f_bresvd;	  /* THIS WONT BE USE IT */
+    fsfilcnt_t f_fresvd;	  /* THIS WONT BE USE IT */
+   struct statvfs svfs;
+};
+
 static char	*getmntpt(const char *);
-static void	 addstat(struct statvfs *, const struct statvfs *);
-static void	 prtstat(const struct statvfs *, int);
+static void	 addstat(struct mntinfo *, const struct statvfs *);
+static void	 prtstat(struct mntinfo *, int);
 static int	 selected(const char *, size_t);
 static void	 maketypelist(char *);
-static size_t	 regetmntinfo(struct statvfs **, size_t);
 __dead static void usage(void);
 static void	 prthumanval(int64_t, int);
 static void	 prthuman(const struct statvfs *, int64_t, int64_t);
 
+static size_t    regetmntinfo(struct mntinfo **, size_t);
+static int        getmntinfo(struct mntinfo **);
+static int        checkvfsselected(char *);
+static int checkvfsname(const char *, const char **, int );
+
 static int	 aflag, cflag, fflag, gflag, hflag, iflag, lflag;
-static int	 Mflag, Nflag, nflag, Pflag, qflag, Wflag;
+static int	 Nflag, nflag, Pflag, Wflag;
 static long	 usize;
 static char	**typelist;
 static size_t	 mntcount;
@@ -87,15 +123,14 @@
 static int blksize_width = WIDTH_BLKSIZE;
 
 static int fudgeunits = 0;
-
-#define	streq(a, b)	(strcmp((a), (b)) == 0)
-#define	warnq(args)	do { if (!qflag) warnx args; } while (0)
+static int        skipvfs_l, skipvfs_t;
+static const char **vfslist_l, **vfslist_t;
 
 int
 main(int argc, char *argv[])
 {
 	struct stat stbuf;
-	struct statvfs *mntbuf, totals;
+	struct mntinfo *mntbuf, totals;
 	int ch, maxwidth, width;
 	size_t i;
 	char *mntpt;
@@ -103,9 +138,7 @@
 	setprogname(argv[0]);
 	(void)setlocale(LC_ALL, "");
 
-	(void)strpct_round(STRPCT_RAZ);	/* POSIX: %'s round away from zero */
-
-	while ((ch = getopt(argc, argv, "abcfGgHhiklMmNnPqt:W")) != -1)
+	while ((ch = getopt(argc, argv, "abcfGgHhiklmNnPt:W")) != -1)
 		switch (ch) {
 		case 'a':
 			aflag = 1;
@@ -144,9 +177,6 @@
 		case 'l':
 			lflag = 1;
 			break;
-		case 'M':
-			Mflag = 1;
-			break;
 		case 'm':
 			hflag = 0;
 			usize = 1024 * 1024;
@@ -160,9 +190,6 @@
 		case 'P':
 			Pflag = 1;
 			break;
-		case 'q':
-			qflag = 1;
-			break;
 		case 'W':
 			Wflag = 1;
 			break;
@@ -200,7 +227,7 @@
 	argc -= optind;
 	argv += optind;
 
-	mntcount = getmntinfo(&mntbuf, MNT_NOWAIT);
+	mntcount = getmntinfo(&mntbuf);
 	if (mntcount == 0)
 		err(EXIT_FAILURE,
 		    "retrieving information on mounted file systems");
@@ -214,8 +241,7 @@
 		for (/*EMPTY*/; *argv != NULL; argv++) {
 			if (stat(*argv, &stbuf) < 0) {
 				if ((mntpt = getmntpt(*argv)) == 0) {
-					if (!qflag)
-						warn("%s", *argv);
+					warn("%s", *argv);
 					continue;
 				}
 			} else if (S_ISBLK(stbuf.st_mode)) {
@@ -223,30 +249,25 @@
 					mntpt = *argv;
 			} else
 				mntpt = *argv;
-
 			/*
 			 * Statfs does not take a `wait' flag, so we cannot
 			 * implement nflag here.
 			 */
-			if (!statvfs(mntpt, &mntbuf[mntcount])) {
-				if (Mflag &&
-				    !streq(mntpt, mntbuf[mntcount].f_mntonname))
-					warnq(("%s is not a mount point",
-					    mntpt));
-				else if (lflag &&
+			if (!statvfs(mntpt, &mntbuf[mntcount].svfs))
+				if (lflag &&
 				    (mntbuf[mntcount].f_flag & MNT_LOCAL) == 0)
-					warnq(("Warning: %s is not a local %s",
-					    *argv, "file system"));
+					warnx("Warning: %s is not a local %s",
+					    *argv, "file system");
 				else if
 				    (!selected(mntbuf[mntcount].f_fstypename,
 					sizeof(mntbuf[mntcount].f_fstypename)))
-					warnq(("Warning: %s mounted as a %s %s",
+					warnx("Warning: %s mounted as a %s %s",
 					    *argv,
 					    mntbuf[mntcount].f_fstypename,
-					    "file system"));
+					    "file system");
 				else
 					++mntcount;
-			} else if (!qflag)
+			else
 				warn("%s", *argv);
 		}
 	}
@@ -270,7 +291,7 @@
 		if (width > maxwidth)
 			maxwidth = width;
 		if (cflag)
-			addstat(&totals, &mntbuf[i]);
+			addstat(&totals, &mntbuf[i].svfs);
 	}
 
 	if (cflag == 0 || fflag == 0)
@@ -288,13 +309,17 @@
 getmntpt(const char *name)
 {
 	size_t count, i;
-	struct statvfs *mntbuf;
-
-	count = getmntinfo(&mntbuf, MNT_NOWAIT);
+	struct mntinfo *mntbuf;
+	
+	count = getmntinfo(&mntbuf);
 	if (count == 0)
 		err(EXIT_FAILURE, "Can't get mount information");
 	for (i = 0; i < count; i++) {
-		if (streq(mntbuf[i].f_mntfromname, name))
+		#ifdef __linux__
+		mntbuf->f_mntfromname=0;
+		mntbuf->f_mntonname=0;
+		#endif
+		if (!strcmp(mntbuf[i].f_mntfromname, name))
 			return mntbuf[i].f_mntonname;
 	}
 	return 0;
@@ -359,14 +384,14 @@
  * filesystem types not in ``fsmask'' and possibly re-stating to get
  * current (not cached) info.  Returns the new count of valid statvfs bufs.
  */
-static size_t
-regetmntinfo(struct statvfs **mntbufp, size_t count)
+size_t
+regetmntinfo(struct mntinfo **mntbufp, size_t count)
 {
 	size_t i, j;
-	struct statvfs *mntbuf;
+	struct mntinfo *mntbuf;
 
 	if (!lflag && typelist == NULL && aflag)
-		return nflag ? count : (size_t)getmntinfo(mntbufp, MNT_WAIT);
+		return nflag ? count : (size_t)getmntinfo(&mntbuf);
 
 	mntbuf = *mntbufp;
 	j = 0;
@@ -381,14 +406,14 @@
 		if (nflag)
 			mntbuf[j] = mntbuf[i];
 		else {
-			struct statvfs layerbuf = mntbuf[i];
-			(void)statvfs(mntbuf[i].f_mntonname, &mntbuf[j]);
+			struct mntinfo layerbuf = mntbuf[i];
+			(void)statvfs(mntbuf[i].f_mntonname, &mntbuf[j].svfs);
 			/*
 			 * If the FS name changed, then new data is for
 			 * a different layer and we don't want it.
 			 */
-			if (memcmp(layerbuf.f_mntfromname,
-			    mntbuf[j].f_mntfromname, MNAMELEN))
+			if (strcmp(layerbuf.f_mntfromname,
+			    mntbuf[j].f_mntfromname))
 				mntbuf[j] = layerbuf;
 		}
 		j++;
@@ -427,42 +452,45 @@
 	    (int64_t)(num) * (int64_t)((fsbs) / (bs)))
 
 static void
-addstat(struct statvfs *totalfsp, const struct statvfs *sfsp)
+addstat(struct mntinfo *totalfsp, const struct statvfs *sfsp)
 {
+	struct mntinfo *totalsfp, *sfps;
 	uint64_t frsize;
 
 	frsize = sfsp->f_frsize / totalfsp->f_frsize;
 	totalfsp->f_blocks += sfsp->f_blocks * frsize;
 	totalfsp->f_bfree += sfsp->f_bfree * frsize;
 	totalfsp->f_bavail += sfsp->f_bavail * frsize;
-	totalfsp->f_bresvd += sfsp->f_bresvd * frsize;
+	totalsfp->f_bresvd += sfps->f_bresvd * frsize;
 	totalfsp->f_files += sfsp->f_files;
 	totalfsp->f_ffree += sfsp->f_ffree;
 	totalfsp->f_favail += sfsp->f_favail;
-	totalfsp->f_fresvd += sfsp->f_fresvd;
+	totalsfp->f_fresvd += sfps->f_fresvd;
 }
 
 /*
  * Print out status about a filesystem.
  */
 static void
-prtstat(const struct statvfs *sfsp, int maxwidth)
+prtstat(struct mntinfo *sfps, int maxwidth)
 {
-	static long blocksize;
+	struct statvfs *sfsp = &sfps->svfs;
+
+	static long blocksize = 1024;
 	static int headerlen, timesthrough;
 	static const char *header;
 	static const char full[] = "100";
 	static const char empty[] = "  0";
-	int64_t used, availblks, inodes;
-	int64_t bavail;
+	uint64_t used, availblks, inodes;
+	uint64_t bavail;
 	char pb[64];
-	char mntfromname[sizeof(sfsp->f_mntfromname) + 10];
+	char mntfromname[128];
 
-	if (Wflag && sfsp->f_mntfromlabel[0]) {
+	if (Wflag && sfps->f_mntfromlabel[0]) {
 		snprintf(mntfromname, sizeof(mntfromname), "NAME=%s",
-		    sfsp->f_mntfromlabel);
+		    sfps->f_mntfromlabel);
 	} else {
-		strlcpy(mntfromname, sfsp->f_mntfromname, sizeof(mntfromname));
+		snprintf(mntfromname, sizeof(mntfromname), "%s", sfps->f_mntfromname);
 	}
 
 	if (gflag) {
@@ -476,7 +504,7 @@
 		 *
 		 */
 		(void)printf("%10s (%-12s): %7ld block size %12ld frag size\n",
-		    sfsp->f_mntonname, mntfromname,
+		    sfps->f_mntonname, mntfromname,
 		    sfsp->f_bsize,	/* On UFS/FFS systems this is
 					 * also called the "optimal
 					 * transfer block size" but it
@@ -488,20 +516,20 @@
 					 * block size" is the frag
 					 * size.
 					 */
-		(void)printf("%10" PRId64 " total blocks %10" PRId64
+		(void)printf("%10" PRIu64 " total blocks %10" PRIu64
 		    " free blocks  %10" PRId64 " available\n",
 		    (uint64_t)sfsp->f_blocks, (uint64_t)sfsp->f_bfree,
 		    (uint64_t)sfsp->f_bavail);
-		(void)printf("%10" PRId64 " total files  %10" PRId64
+		(void)printf("%10" PRIu64 " total files  %10" PRIu64
 		    " free files %12lx filesys id\n",
-		    (uint64_t)sfsp->f_files, (uint64_t)sfsp->f_ffree,
+		    (uint64_t)sfsp->f_ffree, (uint64_t)sfsp->f_files,
 		    sfsp->f_fsid);
 		(void)printf("%10s fstype  %#15lx flag  %17ld filename "
-		    "length\n", sfsp->f_fstypename, sfsp->f_flag,
+		    "length\n", sfps->f_fstypename, sfsp->f_flag,
 		    sfsp->f_namemax);
-		(void)printf("%10lu owner %17" PRId64 " syncwrites %12" PRId64
-		    " asyncwrites\n\n", (unsigned long)sfsp->f_owner,
-		    sfsp->f_syncwrites, sfsp->f_asyncwrites);
+		(void)printf("%10lu owner %17" PRIu64 " syncwrites %12" PRId64
+		    " asyncwrites\n\n", (unsigned long)sfps->f_owner,
+		    sfps->f_syncwrites, sfps->f_asyncwrites);
 
 		/*
 		 * a concession by the structured programming police to the
@@ -572,8 +600,8 @@
 			(void)printf(" Mounted on\n");
 		}
 	}
-	used = sfsp->f_blocks - sfsp->f_bfree;
-	bavail = sfsp->f_bfree - sfsp->f_bresvd;
+	used = sfps->f_blocks;
+	bavail = sfps->f_bavail;
 	availblks = bavail + used;
 	if (Pflag) {
 		assert(hflag == 0);
@@ -584,12 +612,12 @@
 		 * <file system root>
 		 */
 		(void)printf("%s %" PRId64 " %" PRId64 " %" PRId64 " %s%% %s\n",
-		    mntfromname,
+		    sfps->f_mntfromname,
 		    fsbtoblk(sfsp->f_blocks, sfsp->f_frsize, blocksize),
 		    fsbtoblk(used, sfsp->f_frsize, blocksize),
 		    fsbtoblk(bavail, sfsp->f_frsize, blocksize),
 		    availblks == 0 ? full : strspct(pb, sizeof(pb), used,
-		    availblks, 0), sfsp->f_mntonname);
+		    availblks, 0), sfps->f_mntonname);
 		/*
 		 * another concession by the structured programming police to
 		 * the indentation police....
@@ -609,7 +637,7 @@
 			    sfsp->f_frsize, blocksize));
 
 		if (mntcount != 1)
-			(void)printf(" %s\n", sfsp->f_mntonname);
+			(void)printf(" %s\n", sfps->f_mntonname);
 		else
 			(void)printf("\n");
 		return;
@@ -637,7 +665,108 @@
 		    inodes == 0 ? (used == 0 ? empty : full) :
 		    strspct(pb, sizeof(pb), used, inodes, 0));
 	}
-	(void)printf(" %s\n", sfsp->f_mntonname);
+	(void)printf(" %s\n", sfps->f_mntonname);
+}
+
+int
+getmntinfo(struct mntinfo **mntbuf)
+{
+    struct mntinfo *list = NULL;
+    struct mntinfo *cur;
+    struct mntent *ent;
+    struct statvfs sv;
+    struct stat st;
+    FILE *fp;
+    int n = 0;
+
+    fp = setmntent("/proc/self/mounts", "r");
+    if (fp == NULL)
+        err(1, "setmntent");
+
+    while ((ent = getmntent(fp)) != NULL) {
+
+        if (hasmntopt(ent, MNTTYPE_IGNORE))
+            continue;
+
+        if (vfslist_l || vfslist_t) {
+            if (checkvfsselected(ent->mnt_type) != 0)
+                continue;
+        }
+
+        if (statvfs(ent->mnt_dir, &sv) == -1)
+            continue;
+
+        if (stat(ent->mnt_dir, &st) == -1)
+            continue;
+
+        list = realloc(list, (n + 1) * sizeof(*list));
+        if (!list)
+            err(1, "realloc");
+
+        cur = &list[n];
+
+        memset(cur, 0, sizeof(*cur));
+
+        cur->f_fstypename = strdup(ent->mnt_type);
+        cur->f_mntfromname = strdup(ent->mnt_fsname);
+        cur->f_mntonname = strdup(ent->mnt_dir);
+        cur->f_opts = strdup(ent->mnt_opts);
+
+        cur->f_blocks = sv.f_blocks;
+        cur->f_bfree  = sv.f_bfree;
+        cur->f_bavail = sv.f_bavail;
+        cur->f_files  = sv.f_files;
+        cur->f_ffree  = sv.f_ffree;
+        cur->f_bsize  = sv.f_bsize;
+        cur->f_frsize = sv.f_frsize;
+        cur->f_flag   = sv.f_flag;
+        cur->f_namemax = sv.f_namemax;
+
+        /* stat */
+        cur->f_dev = st.st_dev;
+
+        cur->f_selected = 1;
+        n++;
+    }
+
+    endmntent(fp);
+
+    *mntbuf = list;
+    return n;
+}
+
+int
+checkvfsname(const char *vfsname, const char **vfslist, int skip)
+{
+
+	if (vfslist == NULL)
+		return (0);
+	while (*vfslist != NULL) {
+		if (strcmp(vfsname, *vfslist) == 0)
+			return (skip);
+		++vfslist;
+	}
+	return (!skip);
+}
+
+int
+checkvfsselected(char *fstypename)
+{
+	int result;
+
+	if (vfslist_t) {
+		/* if -t option used then select passed types */
+		result = checkvfsname(fstypename, vfslist_t, skipvfs_t);
+		if (vfslist_l) {
+			/* if -l option then adjust selection */
+			if (checkvfsname(fstypename, vfslist_l, skipvfs_l) == skipvfs_t)
+				result = skipvfs_t;
+		}
+	} else {
+		/* no -t option then -l decides */
+		result = checkvfsname(fstypename, vfslist_l, skipvfs_l);
+	}
+	return (result);
 }
 
 static void
@@ -645,9 +774,10 @@
 {
 
 	(void)fprintf(stderr,
-	    "Usage: %s [-aclMnqW] [-G|-bkP|-bfgHhikmN] [-t type] [file | "
+	    "Usage: %s [-aclnW] [-G|-bkP|-bfgHhikmN] [-t type] [file | "
 	    "file_system]...\n",
 	    getprogname());
 	exit(1);
 	/* NOTREACHED */
 }
+
