--- src/tsort/tsort.c	2026-04-01 11:50:38.409531071 -0500
+++ src/tsort/tsort.c	2026-03-29 10:33:00.917843993 -0500
@@ -36,7 +36,7 @@
 #include "nbtool_config.h"
 #endif
 
 #include <sys/cdefs.h>
 #if !defined(lint)
 __COPYRIGHT("@(#) Copyright (c) 1989, 1993, 1994\
  The Regents of the University of California.  All rights reserved.");
@@ -48,7 +48,6 @@
 
 #include <sys/types.h>
 #include <ctype.h>
-#include <db.h>
 #include <err.h>
 #include <errno.h>
 #include <fcntl.h>
@@ -56,6 +55,7 @@
 #include <stdlib.h>
 #include <string.h>
 #include <unistd.h>
 #if !HAVE_NBTOOL_CONFIG_H
 #include <util.h>
 #endif
@@ -100,14 +100,90 @@
 	size_t b_bsize;
 } BUF;
 
-static DB *db;
+#define HASH_CHUNK_SIZE 64
+#define HASH_BUCKET_COUNT 1024
+
+struct hash_elem {
+	NODE *elem;
+	struct hash_elem *next;
+};
+
+struct hash_chunk {
+	struct hash_elem elems[HASH_CHUNK_SIZE];
+	struct hash_chunk *next;
+};
+
+struct hash {
+	struct hash_elem **elems;
+	struct hash_chunk *chunks;
+	struct hash_elem *top;
+};
+
+static void hash_init(struct hash *h) {
+	h->chunks = NULL;
+	h->top = NULL;
+	h->elems = calloc(1024, sizeof(struct hash_elem *));
+}
+
+static void hash_destroy(struct hash *h) {
+	for (size_t i = 0; i < HASH_BUCKET_COUNT; ++i) {
+		struct hash_elem *e = h->elems[i];
+		while (e) {
+			free(e->elem->n_arcs);
+			free(e->elem);
+			e = e->next;
+		}
+	}
+	free(h->elems);
+	while (h->chunks) {
+		struct hash_chunk *c = h->chunks;
+		h->chunks = h->chunks->next;
+		free(c);
+	}
+}
+
+static size_t hash_key(char *key) {
+	size_t h = 5381;
+	for (size_t i = 0, k; (k = key[i]); ++i)
+		h = ((h << 5) + h) ^ k;
+	return h;
+}
+
+static NODE *hash_find(struct hash *h, char *key) {
+	size_t hash = hash_key(key) & (HASH_BUCKET_COUNT - 1);
+	for (struct hash_elem *c = h->elems[hash]; c; c = c->next) {
+		if (!strcmp(key, c->elem->n_name))
+			return c->elem;
+	}
+	return NULL;
+}
+
+static struct hash_elem *hash_insert(struct hash *h, char *key) {
+	size_t hash = hash_key(key) & (HASH_BUCKET_COUNT - 1);
+	if (!h->top) {
+		struct hash_chunk *c = calloc(1, sizeof(struct hash_chunk));
+		c->next = h->chunks;
+		h->chunks = c;
+		for (size_t i = 0; i < (HASH_CHUNK_SIZE - 1); ++i) 
+			c->elems[i].next = &c->elems[i + 1];
+		c->elems[HASH_CHUNK_SIZE - 1].next = h->top;
+		h->top = c->elems;
+	}
+	struct hash_elem *hc = h->top;
+	h->top = h->top->next;
+	hc->next = h->elems[hash];
+	h->elems[hash] = hc;
+	return hc;
+}
+
+static struct hash db;
 static NODE *graph, **cycle_buf, **longest_cycle;
 static int debug, longest, quiet, reverse;
 
-static void	 add_arc(const char *, const char *);
+static void	 add_arc(char *, char *);
 static void	 clear_cycle(void);
 static size_t	 find_cycle(NODE *, NODE *, size_t, size_t);
-static NODE	*get_node(const char *);
+static NODE	*get_node(char *);
 static void	 remove_node(NODE *);
 static void	 tsort(void);
 __dead static void	 usage(void);
@@ -160,6 +236,7 @@
 	for (b = bufs, n = 2; n-- > 0; b++)
 		b->b_buf = erealloc(NULL, b->b_bsize = 1024);
 
+	hash_init(&db);
 	/* parse input and build the graph */
 	for (n = 0, c = getc(fp);;) {
 		while (c != EOF && isspace(c))
@@ -193,6 +270,7 @@
 
 	/* do the sort */
 	tsort();
+	hash_destroy(&db);
 	return EXIT_SUCCESS;
 }
 
@@ -201,7 +279,7 @@
  * the graph, then add them.
  */
 static void
-add_arc(const char *s1, const char *s2)
+add_arc(char *s1, char *s2)
 {
 	NODE *n1;
 	NODE *n2;
@@ -236,37 +314,25 @@
 
 /* Find a node in the graph (insert if not found) and return a pointer to it. */
 static NODE *
-get_node(const char *name)
+get_node(char *name)
 {
-	DBT data, key;
-	NODE *n;
-
-	if (db == NULL &&
-	    (db = dbopen(NULL, O_RDWR, 0, DB_HASH, NULL)) == NULL)
-		err(EXIT_FAILURE, "db: %s", name);
-
-	key.data = __UNCONST(name);
-	key.size = strlen(name) + 1;
+	NODE *n = hash_find(&db, name);
+	size_t nlen;
 
-	switch ((*db->get)(db, &key, &data, 0)) {
-	case 0:
-		(void)memmove(&n, data.data, sizeof(n));
+	if (n)
 		return n;
-	case 1:
-		break;
-	default:
-	case -1:
-		err(EXIT_FAILURE, "db: %s", name);
-	}
 
-	n = emalloc(sizeof(NODE) + key.size);
+	nlen = strlen(name) + 1;
+
+	if ((n = emalloc(sizeof(NODE) + nlen)) == NULL)
+		err(1, NULL);
 
 	n->n_narcs = 0;
 	n->n_arcsize = 0;
 	n->n_arcs = NULL;
 	n->n_refcnt = 0;
 	n->n_flags = 0;
-	(void)memmove(n->n_name, name, key.size);
+	(void)memcpy(n->n_name, name, nlen);
 
 	/* Add to linked list. */
 	if ((n->n_next = graph) != NULL)
@@ -275,10 +341,7 @@
 	graph = n;
 
 	/* Add to hash table. */
-	data.data = &n;
-	data.size = sizeof(n);
-	if ((*db->put)(db, &key, &data, 0))
-		err(EXIT_FAILURE, "db: %s", name);
+	hash_insert(&db, name)->elem = n;
 	return n;
 }
 
